### ISSUE TITLE
torch.no_grad(fn) 

### ISSUE THREAD

**Issue by PaulScemama** [pytorch/pytorch#107036]
### 🚀 The feature, motivation and pitch

I think it *could* be nice to have the following:

```python
import torch

def no_grad(fn: Callable) -> Callable:
    @torch.no_grad()
    def no_grad_fn(*args, **kwargs):
        return fn(*args, **kwargs)
    return no_grad_fn

def square(x):
    return x**2

square_no_grad = no_grad(square)

if __name__ == "__main__":
    x = torch.tensor([1.0, 2.0, 3.0], requires_grad=True)

    print(square(x))
    # tensor([1., 4., 9.], grad_fn=<PowBackward0>)
    print(square_no_grad(x))
    # tensor([1., 4., 9.])
```

It could be nice to have the `no_grad` be a torch function --> `torch.no_grad(fn)`.

Why? Consider a scenario where we have a core function call. Sometimes we want to have it track gradients, but in other contexts we especially don't. Instead of copying the code twice and decorating one with `@torch.no_grad()`, this seems like a cleaner way.

However, I'm not sure there are so many use cases for this? I have one myself but I don't know how common it will be.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by vadimkantorov** [comment]
I think, this already works, but is indeed not very nice because of extra brackets:
`square_no_grad = torch.no_grad()(lambda x: x ** 2)`

**Comment by andreasfloros** [comment]
@PaulScemama I'm afraid this won't work because we want `no_grad` to act both as a decorator and as a context manager.

That being said, I don't think this should be too hard to implement? One way to do it is by implementing `__new__` on `no_grad` (https://github.com/pytorch/pytorch/blob/main/torch/autograd/grad_mode.py#L16C1-L74C42) as follows:

```python
    def __new__(cls, func = None):  # func is optional to allow both no_grad() and no_grad(func)
        if func:  # if we have a func then no_grad should act a decorator
            @no_grad()
            def wrapper(*args, **kwargs):
                return func(*args, **kwargs)
            return wrapper
        return super().__new__(cls)  # no func, fall back to original implementation
```

And here's a little test:

```python
import torch

a = torch.ones(1, requires_grad=True)


def do_something():
    print(2 * a)


with torch.no_grad():
    do_something()  # tensor([2.])

torch.no_grad()(do_something)()  # tensor([2.])

torch.no_grad(do_something)()  # tensor([2.])

do_something()  # tensor([2.], grad_fn=<MulBackward0>)
```

**Comment by PaulScemama** [comment]
@andreasfloros wonderful! Great stuff

**Comment by colesbury** [comment]
To be clear, `@torch.no_grad` already works as both a decorator and context manager. One of the invocations in @andreasfloros  examples already works. It's a bit awkward because the decorator requires parentheses `()` so the current syntax is a bit unwieldy.

```
torch.no_grad()(square)(x)
```

The modifications to allow the simpler syntax `torch.no_grad(square)(x)` seem fine to me. I suspect that would also allow `@torch.no_grad` as a decorator without the extra parentheses, which also seems fine and more typical of Python decorators that don't take any arguments.

**Comment by andreasfloros** [comment]
> To be clear, `@torch.no_grad` already works as both a decorator and context manager. One of the invocations in @andreasfloros examples already works. It's a bit awkward because the decorator requires parentheses `()` so the current syntax is a bit unwieldy.
> 
> ```
> torch.no_grad()(square)(x)
> ```
> 
> The modifications to allow the simpler syntax `torch.no_grad(square)(x)` seem fine to me. I suspect that would also allow `@torch.no_grad` as a decorator without the extra parentheses, which also seems fine and more typical of Python decorators that don't take any arguments.

Yes, with these changes this also works:

```python
@torch.no_grad
def do_something():
    print(2 * a)
```

As you said, I think this is more pythonic and in my opinion makes the API more user friendly. `dataclass` also behaves in a similar way (https://docs.python.org/3/library/dataclasses.html#module-contents).

I think it would be nice to apply the same logic to `enable_grad` and `inference_mode` which have similar implementations.

**Comment by albanD** [comment]
That sounds good! Feel free to send a PR updating this with appropriate testing!
